src/node_http2.cc

```diff
@@ -1872,7 +1872,10 @@ void Http2Session::OnStreamRead(ssize_t nread, const uv_buf_t& buf_) {
     nread = buf.size();
     stream_buf_offset_ = 0;
     stream_buf_ab_.Reset();
-    DecrementCurrentSessionMemory(stream_buf_offset_);
+
+    // We have now fully processed the stream_buf_ input chunk (by moving the
+    // remaining part into buf, which will be accounted for below).
+    DecrementCurrentSessionMemory(stream_buf_.len);
```

Comment:
I don't think this is correct as we only remove the `stream_buf_offset_` amount of data from the buffer. Could this fix your problem https://github.com/nodejs/node/pull/30534?

Reply:
This isn't really a comment on correctness, but I ran the test from this PR locally against master, #30351, and the changes in this PR. master and #30351 yielded `NGHTTP2_ENHANCE_YOUR_CALM`, while the test passed with the changes in this PR.

Reply:
Oh, funny.  I hadn't seen #30351.  I don't think it's the right fix though.  It's worth noting that in the case we set stream_buf_offset_ to be non-zero at https://github.com/nodejs/node/blob/f36331c1bfa4c4c202346b05dc3bd672f653e4df/src/node_http2.cc#L871, we do not call DecrementCurrentSessionMemory() at all (due to the early return) and so all of stream_buf_ is still being counted as session memory.  So the correct thing to do is decrement all of it here.

Reply:
Oh, I got the issue, I guess instead of going back and forth with this memory I think it will be better to separate `IncrementCurrentSessionMemory` calls like
```c++
  if (UNLIKELY(stream_buf_offset_ > 0)) {
    // This is a very unlikely case, and should only happen if the ReadStart()
    // call in OnStreamAfterWrite() immediately provides data. If that does
    // happen, we concatenate the data we received with the already-stored
    // pending input data, slicing off the already processed part.
    size_t pending_len = stream_buf_.len - stream_buf_offset_;
    AllocatedBuffer new_buf = env()->AllocateManaged(pending_len + nread);
    memcpy(new_buf.data(), stream_buf_.base + stream_buf_offset_, pending_len);
    memcpy(new_buf.data() + pending_len, buf.data(), nread);
    IncrementCurrentSessionMemory(nread - stream_buf_offset_);
    buf = std::move(new_buf);
    nread = buf.size();
    stream_buf_offset_ = 0;
    stream_buf_ab_.Reset();
  } else {
    // Shrink to the actual amount of used data.
    buf.Resize(nread);
    IncrementCurrentSessionMemory(nread);
  }
```

Reply:
Yes, I think that should work too...  although it's worth noting that `nread - stream_buf_offset_` could be negative and so we'll be "incrementing" the session memory by a negative number, which is perhaps a bit unintuitive.

Reply:
Will you update this PR and I will drop the first commit from [mine](https://github.com/nodejs/node/pull/30351) or should I update mine and this one will only add a test? I'm fine either way.

Reply:
Hrm.  To be honest I'm not sure which way is better.  It took me a minute to convince myself that `IncrementCurrentSessionMemory(nread - stream_buf_offset_)` was doing the correct thing and I do think its less intuitive to increment by a negative number.  So I do kind of think that my PR as-is is the most obviously-correct way to fix the leak.

So perhaps you can update your PR and we can leave it to @addaleax or whoever owns this code to weigh in on which approach they prefer and we'll merge that one?

Feel free to just grab my test and add it to your PR.  I don't care which PR gets merged as long as we get this bug fixed. 😄

Reply:
> Feel free to just grab my test and add it to your PR. I don't care which PR gets merged as long as we get this bug fixed. :smile:

Yeah, same :) @lundibundi feel free to do that or land this, whichever you